### PR TITLE
feat: add e2e gen ai app starter pack

### PR DESCRIPTION
**Pull request by eliasecchig**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [x] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: core
[the remark this answers is not in the document]
Comment (eliasecchig) [reply core:1]: done thanks!

File: gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        for human_message, ai_message in pairwise(row["messages"]):
+            message = {
+                "human_message": human_message,
+                "ai_message": ai_message,
+                "conversation_history": conversation_history,
+            }
+            conversation_history = conversation_history + [human_message, ai_message]
+            processed_messages.append(message)
+
+    return pd.DataFrame(processed_messages)
```
Comment (holtskinner) [inline gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py:91]: See if there's a way to change this to use dataframe operations, like `df.apply()` could make it more efficient.

File: gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    """
+    logging.info("Executing batch scoring")
+    predicted_messages = []
+    with ThreadPoolExecutor(max_workers) as pool:
+        partial_func = partial(generate_message, callable=callable)
+        for message in tqdm(
+            pool.map(partial_func, messages.iterrows()), total=len(messages)
+        ):
+            predicted_messages.append(message)
+    return pd.DataFrame(predicted_messages)
```
Comment (holtskinner) [inline gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py:205]: See if you can use `df.apply()` or something similar.

File: gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+
+def save_df_to_csv(df: pd.DataFrame, dir_path: str, filename: str) -> None:
```
Comment (holtskinner) [inline gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/eval/utils.py:208]: Is this method used?

File: gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/utils/tracing.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        :param spans: A sequence of spans to export
+        :return: The result of the export operation
+        """
+        for span in spans:
+            span_context = span.get_span_context()
+            trace_id = format(span_context.trace_id, "x")
+            span_id = format(span_context.span_id, "x")
+            span_dict = json.loads(span.to_json())
+
+            span_dict["trace"] = f"projects/{self.project_id}/traces/{trace_id}"
```
Comment (holtskinner) [inline gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/utils/tracing.py:76]: There should be a method on the client to create this path. Something like `client.trace_path()`

File: gemini/sample-apps/e2e-gen-ai-app-starter-pack/.gitignore
[the remark this answers is not in the document]
Comment (eliasecchig) [reply gemini/sample-apps/e2e-gen-ai-app-starter-pack/.gitignore:1]: Sure happy to do it this way.

**Review by holtskinner — APPROVED** [review]

**Pull request merged**